lib/libhfs/libhfs.c

```diff
@@ -2484,25 +2484,31 @@ hfslib_compare_catalog_keys_bc (
 	const void *a,
 	const void *b)
 {
+	unichar_t ac,   bc;
+	uint8_t   apos, bpos;
+
 	if(((const hfs_catalog_key_t*)a)->parent_cnid
 		== ((const hfs_catalog_key_t*)b)->parent_cnid)
 	{
-		if(((const hfs_catalog_key_t*)a)->name.length == 0 &&
-			((const hfs_catalog_key_t*)b)->name.length == 0)
-			return 0;
+		apos = bpos = 0;
+		while((apos < ((const hfs_catalog_key_t*)a)->name.length)
+			&& (bpos < ((const hfs_catalog_key_t*)b)->name.length))
+		{
+			ac = ((const hfs_catalog_key_t*)a)->name.unicode[apos++];
+			bc = ((const hfs_catalog_key_t*)b)->name.unicode[bpos++];
 
-		if(((const hfs_catalog_key_t*)a)->name.length == 0)
-			return -1;
-		if(((const hfs_catalog_key_t*)b)->name.length == 0)
-			return 1;
+			if (ac == bc)
+				continue;
 
-		/* FIXME: This does a byte-per-byte comparison, whereas the HFS spec
-		 * mandates a uint16_t chunk comparison. */
-		return memcmp(((const hfs_catalog_key_t*)a)->name.unicode,
```

Comment:
I see the issue. It looks like you could keep most of the existing function including the memcmp, but capture its result and return the difference of the lengths if nonzero. I'd prefer to keep the diffs against the NetBSD code small where possible to facilitate keeping them in sync, but also memcmp is likely to be vectorized and should be more performant here.